src/w3d/renderer/linegroup.cpp

```diff
@@ -0,0 +1,358 @@
+{
+    if (texture != nullptr) {
+        texture->Add_Ref();
+    }
+    m_texture = texture;
+}
+
+void LineGroupClass::Render(RenderInfoClass &rinfo)
+{
+#ifdef BUILD_WITH_D3D8
+    if (m_lineCount > 0) {
+        m_shader.Set_Cull_Mode(ShaderClass::CULL_MODE_DISABLE);
+
+        float value_255 = 254.0f / 255.0f;
+
+        bool default_white_opaque = m_defaultColor.X > value_255 && m_defaultColor.Y > value_255
+            && m_defaultColor.Z > value_255 && m_defaultAlpha > value_255;
+
+        if (m_diffuseBuffer != nullptr || !default_white_opaque || m_texture == nullptr) {
+            m_shader.Set_Primary_Gradient(ShaderClass::GRADIENT_MODULATE);
+        } else {
+            m_shader.Set_Primary_Gradient(ShaderClass::GRADIENT_DISABLE);
+        }
+
+        if (m_texture != nullptr) {
+            m_shader.Set_Texturing(ShaderClass::TEXTURING_ENABLE);
+        } else {
+            m_shader.Set_Texturing(ShaderClass::TEXTURING_DISABLE);
+        }
+
+        VertexMaterialClass *material = VertexMaterialClass::Get_Preset(VertexMaterialClass::PRELIT_DIFFUSE);
+
+        DX8Wrapper::Set_Material(material);
+        DX8Wrapper::Set_Shader(m_shader);
+        DX8Wrapper::Set_Texture(0, m_texture);
+
+        if (material != nullptr) {
+            material->Release_Ref();
+        }
+
+        captainslog_assert(m_startLoc != nullptr && m_startLoc->Get_Array());
+        captainslog_assert(m_endLoc != nullptr && m_endLoc->Get_Array());
+
+        const bool sort = m_shader.Get_Dst_Blend_Func() != ShaderClass::DSTBLEND_ZERO
+            && m_shader.Get_Alpha_Test() == ShaderClass::ALPHATEST_DISABLE && W3D::Is_Sorting_Enabled();
+
+        static Vector3 _offset_a(GameMath::Cos(DEG_TO_RAD(90)), GameMath::Sin(DEG_TO_RAD(90)), 0);
+        static Vector3 _offset_b(GameMath::Cos(DEG_TO_RAD(210)), GameMath::Sin(DEG_TO_RAD(210)), 0);
+        static Vector3 _offset_c(GameMath::Cos(DEG_TO_RAD(330)), GameMath::Sin(DEG_TO_RAD(330)), 0);
+        static Vector3 _offset[3];
+
+        _offset[0].Set(_offset_a);
+        _offset[1].Set(_offset_b);
+        _offset[2].Set(_offset_c);
+
+        Matrix4 old_view;
+        DX8Wrapper::Get_Transform(D3DTS_VIEW, old_view);
+
+        Matrix4 identity(true);
+        DX8Wrapper::Set_Transform(D3DTS_WORLD, identity);
+
+        if (Get_Flag(TRANSFORM)) {
+            Matrix3D model_view;
+
+            model_view = rinfo.m_camera.Get_Transform();
+            model_view.Set_Translation(Vector3(0, 0, 0));
+            model_view.Get_Orthogonal_Inverse(model_view);
+
+            for (int i = 0; i < 3; ++i) {
+                Matrix3D::Transform_Vector(model_view, _offset[i], &_offset[i]);
+            }
+        } else {
+            DX8Wrapper::Set_Transform(D3DTS_VIEW, identity);
+        }
+
+        int polygon_count = 0;
+        int index_count = 0;
+        int vertex_count = 0;
+
+        if (m_mode == TETRAHEDRON) {
+            polygon_count = 4 * m_lineCount;
```

Comment:
It would be useful to extract out the constants to somewhere to ease understanding:

```cpp
constexpr int32_t TETRAHEADRON_NUM_POLYGONS = 4;
constexpr int32_t TETRAHEADRON_NUM_VERTEXS = 4;

constexpr int32_t PRISM_NUM_POLYGONS = 8;
constexpr int32_t PRISM_NUM_VERTEXS = 6;
```
Then the other constants in this function would make more sense such as at L194